### DISCUSSION TITLE
Dependency defined on startup event

### DISCUSSION THREAD

**Discussion by gygabyte017** [fastapi/fastapi#6614] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
@app.on_event("startup")
async def startup_event():
    something = await func()
    myobj = MyObj(something)

###

@router.get("")
async def index():
  # need myobj instance here
```

### Description

I need to execute some async initialization on startup creating an instance of `myobj`, and then I need this instance as a dependency in routers endpoint.

### Operating System

Windows

### FastAPI Version

0.68.1

### Python Version

3.8.8

### Additional Context

I need to create an instance of a "master" object asynchronously during the startup of FastApi, and then to get this exact instance in routers endpoint, but I cannot figure out how to pass it as a dependency.

(I'm aware that if I have multiple workers each worker will get a different "master" object but that is ok, the main point here is that within a single worker an instance is created after async operation on startup, and then reused in routers endpoint.)

**Comment by insomnes** [comment]
You can try to use some kind of singleton implementation for that object. And return it's instance with dependency injection:
https://fastapi.tiangolo.com/tutorial/dependencies/

**Comment by gygabyte017** [comment]
Thanks it works, but it seems to me a bit "hacky". Perhaps dependencies are not the correct tool here?

**Comment by insomnes** [comment]
If you use singleton, you can initiate your object inside handler function. 
But it is really easier to test and change your code, if you use Dependency Injection system.
In any way it's up to you.

**Comment by STeveShary** [comment]
@gygabyte017 I think if you look at the use of a static class variable, you should be able to accomplish your goal.

If we want to have a singleton object that exists once, we can do this with a variable that is static in a class:

```python3

class SingletonClass:

  _singleton_object = None

  @staticmethod
  async def _init_Singleton():
    pass


  @staticmethod
  async def get_singleton():
    if not SingletonClass._singleton_object:
      await  SingletonClass._init_singleton()
    return SingletonClass._singleton_object

```

Then we can Modify your code to be:

```python3
@app.on_event("startup")
async def startup_event():
    something = await SingletonClass.get_singleton()

###

@router.get("")
async def index():
 SingletonClass.get_singleton()
```

This will have one instance of the object that can be references from any place.

**Comment by gygabyte017** [comment]
Thanks to both of you!

**Comment by adriangb** [comment, marked as answer]
Another alternative is to store it in the app state:

```python
from fastapi import FastAPI, Depends, Request


class MyObj:
    def __init__(self, something: int) -> None:
        self.something = something


async def on_startup() -> None:
    app.state.myobj = MyObj(1)


app = FastAPI(on_startup=[on_startup])


def get_myobj(request: Request) -> MyObj:
    return request.app.state.myobj


@app.get("/")
async def index(myobj: MyObj = Depends(get_myobj)) -> None:
    assert isinstance(myobj, MyObj)
    assert myobj.something == 1


from fastapi.testclient import TestClient


with TestClient(app) as client:
    resp = client.get("/")
    assert resp.status_code == 200
```

Or, more in line with traditional dependency injection if you imagine that `dependency_overrides` is a "container":

```python
from fastapi import FastAPI, Depends


class MyObj:
    def __init__(self, something: int) -> None:
        self.something = something


async def on_startup() -> None:
    myobj = MyObj(1)
    app.dependency_overrides[MyObj] = lambda: myobj


app = FastAPI(on_startup=[on_startup])


@app.get("/")
async def index(myobj: MyObj = Depends()) -> None:
    assert isinstance(myobj, MyObj)
    assert myobj.something == 1


from fastapi.testclient import TestClient


with TestClient(app) as client:
    resp = client.get("/")
    assert resp.status_code == 200
```

If this was a more realistic situation, I'd hope this looked like:

```python
from fastapi import FastAPI, Depends


class MyObj:
    def __init__(self, something: int) -> None:
        self.something = something


async def on_startup(container: Container) -> None:
    # get the container or app injected, no need for a reference
    container.bind(MyObj, value=MyObj(1))   # or something like this


app = FastAPI(on_startup=[on_startup])


@app.get("/")
async def index(myobj: MyObj = Depends()) -> None:
    ...
```

**Reply by ogtega** [reply]
Could you expand on the third example?
Not sure what container is in this context and what the equivalent would be with the new context manager annotation for startup events.

**Reply by adriangb** [reply]
It’s made up pseudo code. That API doesn’t exist.

**Reply by ogtega** [reply]
Ahh, didn't read "I'd hope" 😆

**Comment by desilinguist** [comment]
@adriangb thank you so much for the idea about using dependency overrides. It really helped me out!

**Comment by tiangolo** [comment]
Thanks for the discussion here everyone! ☕

And thanks @gygabyte017 for coming back to close the issue.

**Comment by UltimateLobster** [comment]
I made it all the way to this discussion from searching on dependency injection in the events handling. Is that still planned to be released?

**Comment by martimors** [comment]
I think FastAPI should have a way to instantiate (thus cache) dependencies on start up! Currently, we have to write custom logic to verify dependencies right?
